spring-kafka/src/main/java/org/springframework/kafka/support/serializer/JsonDeserializer.java

```diff
@@ -158,13 +158,22 @@ public JsonDeserializer(Class<T> targetType, ObjectMapper objectMapper) {
 	@SuppressWarnings("unchecked")
 	public JsonDeserializer(@Nullable Class<T> targetType, ObjectMapper objectMapper, boolean useHeadersIfPresent) {
 		Assert.notNull(objectMapper, "'objectMapper' must not be null.");
-		Assert.isTrue(targetType != null || useHeadersIfPresent,
-				"'targetType' cannot be null if 'useHeadersIfPresent' is false");
 		this.objectMapper = objectMapper;
 		if (targetType == null) {
-			targetType = (Class<T>) ResolvableType.forClass(getClass()).getSuperType().resolveGeneric(0);
+			this.targetType = (Class<T>) ResolvableType.forClass(getClass()).getSuperType().resolveGeneric(0);
+			if (this.targetType != null) {
+				this.reader = this.objectMapper.readerFor(this.targetType);
+			}
+			else {
+				this.reader = null;
+			}
+		}
+		else {
+			this.targetType = targetType;
+			this.reader = this.objectMapper.readerFor(this.targetType);
```

Comment:
I think we can have this line of code only once: after that `Assert.isTrue()` and within `if...else` around `this.targetType`.
Just a nit-pick do keep DRY principle...